src/rust/engine/src/externs/mod.rs

```diff
@@ -190,57 +171,49 @@ pub fn val_to_log_level(obj: &PyObject) -> Result<log::Level, String> {
 
 /// Link to the Pants docs using the current version of Pants.
 pub fn doc_url(py: Python, slug: &str) -> String {
-  let docutil = py.import("pants.util.docutil").unwrap();
-  docutil
-    .call(py, "doc_url", (slug,), None)
-    .unwrap()
-    .extract(py)
-    .unwrap()
+  let docutil_module = py.import("pants.util.docutil").unwrap();
+  let doc_url_func = docutil_module.getattr("doc_url").unwrap();
+  doc_url_func.call1((slug,)).unwrap().extract().unwrap()
 }
 
 pub fn create_exception(py: Python, msg: String) -> Value {
-  Value::from(PyErr::new::<cpython::exc::Exception, _>(py, msg).instance(py))
-}
-
-pub fn call_method0(py: Python, value: &PyObject, method: &str) -> Result<PyObject, PyErr> {
-  value.call_method(py, method, PyTuple::new(py, &[]), None)
+  Value::new(PyException::new_err(msg).into_py(py))
 }
 
-pub fn call_function<T: AsRef<PyObject>>(func: T, args: &[Value]) -> Result<PyObject, PyErr> {
-  let func: &PyObject = func.as_ref();
-  let arg_handles: Vec<PyObject> = args.iter().map(|v| v.clone().into()).collect();
-  let gil = Python::acquire_gil();
-  let args_tuple = PyTuple::new(gil.python(), &arg_handles);
-  func.call(gil.python(), args_tuple, None)
+pub fn call_function<'py>(func: &'py PyAny, args: &[Value]) -> PyResult<&'py PyAny> {
+  let args: Vec<PyObject> = args.iter().map(|v| v.clone().into()).collect();
+  let args_tuple = PyTuple::new(func.py(), &args);
+  func.call1(args_tuple)
 }
 
-pub fn generator_send(generator: &Value, arg: &Value) -> Result<GeneratorResponse, Failure> {
-  let gil = Python::acquire_gil();
-  let py = gil.python();
+pub fn generator_send(
+  py: Python,
+  generator: &Value,
+  arg: &Value,
+) -> Result<GeneratorResponse, Failure> {
   let selectors = py.import("pants.engine.internals.selectors").unwrap();
-  let response = selectors
-    .call(
-      py,
-      "native_engine_generator_send",
-      (generator as &PyObject, arg as &PyObject),
-      None,
-    )
+  let native_engine_generator_send = selectors.getattr("native_engine_generator_send").unwrap();
+  let response = native_engine_generator_send
+    .call1((generator.to_object(py), arg.to_object(py)))
     .map_err(|py_err| Failure::from_py_err_with_gil(py, py_err))?;
 
-  if let Ok(b) = response.cast_as::<PyGeneratorResponseBreak>(py) {
-    Ok(GeneratorResponse::Break(Value::new(
-      b.val(py).clone_ref(py),
-    )))
-  } else if let Ok(get) = response.cast_as::<PyGeneratorResponseGet>(py) {
+  if let Ok(b) = response.extract::<PyRef<PyGeneratorResponseBreak>>() {
+    Ok(GeneratorResponse::Break(
+      Value::new(b.0.clone_ref(py)),
+      TypeId::new(b.0.clone_ref(py).into_ref(py).get_type()),
```

Comment:
This is a new change for `GeneratorResponse::Break` to store the `TypeId` eagerly. Before, we'd only store the `Value` and then extract the `TypeId` after. But to do that would require getting the GIL again at the callsite, and it was proving really complicated to port. So instead, we now do both things in the same place.